### PR TITLE
Response mode restriction for FAPI code flow

### PR DESCRIPTION
**Pull request by RivinduM**
### Proposed changes in this pull request

According to the FAPI compliance, the authorization server must require the response mode to be jwt if the response type is code[1]. With this PR, the response mode jwt is validated as a mandatory parameter for the code flow if the application is FAPI compliant.

[1] https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server

### Related Issues
https://github.com/wso2/product-is/issues/16513

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeResponseValidator.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    }
+
+    @Override
+    public void validateRequiredParameters(HttpServletRequest request) throws OAuthProblemException {
+
+        super.validateRequiredParameters(request);
+        // FAPI requests require the response_mode to be jwt when the response_type is code.
+        if (request.getAttribute(OAuthConstants.IS_FAPI_CONFORMANT_APP) != null &&
+                Boolean.parseBoolean(request.getAttribute(OAuthConstants.IS_FAPI_CONFORMANT_APP).toString())) {
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeResponseValidator.java:47]: Move the condition to a new method

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthCommonUtil.java [hunk 2]
```diff
@@ -68,4 +71,20 @@ public static void validateContentTypes(HttpServletRequest request) throws OAuth
             throw OAuthUtils.handleBadContentTypeException(String.join(" or ", ALLOWED_CONTENT_TYPES));
         }
     }
+
+    /**
+     * Method to decode the request object jwt.
+     * @param requestObjectJwt request object jwt
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthCommonUtil.java:77]: Formatting issue.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -291,6 +291,15 @@ public Response authorize(@Context HttpServletRequest request, @Context HttpServ
...
         }
 
+        // Add the isFAPIConformant as a request attribute to retrieve from response validators to avoid the cyclic
+        // dependency in accessing the OAuth2Util.isFapiConformantApp() method.
+        try {
+            request.setAttribute(OAuthConstants.IS_FAPI_CONFORMANT_APP,
+                    OAuth2Util.isFapiConformantApp(oAuthMessage.getClientId()));
+        } catch (IdentityOAuth2Exception e) {
+            EndpointUtil.triggerOnAuthzRequestException(e, request);
+            throw new InvalidRequestException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST, e);
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:301]: Change to a server error

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -291,6 +291,15 @@ public Response authorize(@Context HttpServletRequest request, @Context HttpServ
             return handleIdentityException(request, e);
         }
 
+        // Add the isFAPIConformant as a request attribute to retrieve from response validators to avoid the cyclic
+        // dependency in accessing the OAuth2Util.isFapiConformantApp() method.
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:295]: Multi line comment

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java [hunk 5]
```diff
@@ -273,4 +275,15 @@ private void validateRequestObject(OAuthAuthzRequest oAuthAuthzRequest) throws P
...
     }
+
+    private void setIsFAPIConformantApp(HttpServletRequest request, String clientId) throws ParClientException {
+        // Add the isFAPIConformant as a request attribute to retrieve from response validators to avoid the cyclic
+        // dependency in accessing the OAuth2Util.isFapiConformantApp() method.
+        try {
+            request.setAttribute(OAuthConstants.IS_FAPI_CONFORMANT_APP, OAuth2Util.isFapiConformantApp(clientId));
+        } catch (IdentityOAuth2Exception e) {
+            EndpointUtil.triggerOnAuthzRequestException(e, request);
+            throw new ParClientException(OAuth2ErrorCodes.INVALID_REQUEST, e.getMessage(), e);
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java:286]: Is there any reason to use it as client exception

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 3 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:301]: Move to a private method

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeResponseValidator.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.OAuth20Params.RESPONSE_MODE;
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.ResponseModes.JWT;
+
+/**
+ * Validator for code flow requests.
+ */
+public class CodeResponseValidator extends CodeValidator {
+
+    public CodeResponseValidator() {
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeResponseValidator.java:37]: No need to explicitly have default constructor

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java [hunk 7]
```diff
@@ -273,4 +275,15 @@ private void validateRequestObject(OAuthAuthzRequest oAuthAuthzRequest) throws P
             throw new ParClientException(e.getErrorCode(), e.getMessage(), e);
         }
     }
+
+    private void setIsFAPIConformantApp(HttpServletRequest request, String clientId) throws ParClientException {
+        // Add the isFAPIConformant as a request attribute to retrieve from response validators to avoid the cyclic
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java:280]: Add new line

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeResponseValidator.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Override
+    public void validateRequiredParameters(HttpServletRequest request) throws OAuthProblemException {
+
+        super.validateRequiredParameters(request);
+        // FAPI requests require the response_mode to be jwt when the response_type is code.
+        if (request.getAttribute(OAuthConstants.IS_FAPI_CONFORMANT_APP) != null &&
+                Boolean.parseBoolean(request.getAttribute(OAuthConstants.IS_FAPI_CONFORMANT_APP).toString())) {
+            String responseMode = request.getParameter(RESPONSE_MODE);
+            if (StringUtils.isNotBlank(request.getParameter(OAuthConstants.OAuth20Params.REQUEST))) {
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeResponseValidator.java:49]: Check whether we can move after req obj validation

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java [hunk 9]
```diff
@@ -232,6 +233,7 @@ private void validateAuthzRequest(HttpServletRequest request) throws ParCoreExce
 
         try {
             OAuthAuthzRequest oAuthAuthzRequest = getOAuthAuthzRequest(request);
+            setIsFAPIConformantApp(request, oAuthAuthzRequest.getClientId());
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java:236]: check whether we can move this after req obj validation or before building oauth request

**Comment by RivinduM** [comment]
Closing the PR since there are considerable changes suggested in the review

**Pull request closed without merging**